File: LeetCode/Explore/TopInterview/Easy/LinkedList/Blind75/206_reverse_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
  
public:
  ListNode* reverseList(ListNode* head, ListNode *pre = nullptr) {
    // if we reached the end, then the previous node must be the head of the reversed list
    if (!head) {
      return pre;
    }
    ListNode *next = head->next; // Remember the next node
    head->next = pre; // Reverse the current node
    return reverseList(next, head); // Reverse the remainder of the list
  }
};

File: LeetCode/Explore/TopInterview/Amazon/LinkedList/206_reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
          return head
        
        # next = head.next
        # head.next = pre
        # pre = head
        # head = next
        # return self.reverseList(next, head)
        
        h = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return h
